resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/IFile.java

```diff
@@ -357,6 +358,86 @@ public interface IFile extends IResource, IEncodedStorage, IAdaptable {
+	 *                {@link IResource#setDerived(boolean)} immediately after
+	 *                creating the resource
+	 * @param monitor a progress monitor, or <code>null</code> if progress reporting
+	 *                is not desired
+	 * @throws CoreException if this method fails or is canceled.
+	 * @since 3.21
+	 */
+	public default void create(byte[] content, boolean force, boolean derived, IProgressMonitor monitor)
+			throws CoreException {
+		int createFlags = (force ? IResource.FORCE : IResource.NONE) | (derived ? IResource.DERIVED : IResource.NONE);
+		create(content, createFlags, monitor);
+	}
+
+	/**
+	 * Creates the file with the given content as byte array. Same as calling
+	 * {@link #create(InputStream, int, IProgressMonitor)} and a
+	 * {@code new ByteArrayInputStream(content)} as {@code InputStream}. This method
+	 * is preferably over the streaming API when the content is available as byte
+	 * array anyway.
+	 *
+	 * @param content     the content as byte array
+	 * @param createFlags bit-wise or of flag constants ({@link IResource#FORCE},
+	 *                    {@link IResource#DERIVED}, and
+	 *                    {@link IResource#TEAM_PRIVATE})
+	 * @param monitor     a progress monitor, or <code>null</code> if progress
+	 *                    reporting is not desired
+	 * @throws CoreException if this method fails or is canceled.
+	 * @since 3.21
+	 */
+	public default void create(byte[] content, int createFlags, IProgressMonitor monitor) throws CoreException {
+		create(new ByteArrayInputStream(content), createFlags, monitor);
+	}
+
+	/**
+	 * Creates the file and sets the file content. If the file already exists in
+	 * workspace its content is updated. Shortcut for calling
+	 * {@link #setContents(byte[], boolean, boolean, IProgressMonitor)} or
+	 * {@link #create(byte[], boolean, boolean, IProgressMonitor)} if the file does
+	 * not {@link #exists()}.
+	 *
+	 * @param content     the new content bytes. Must not be null.
+	 * @param force       a flag controlling how to deal with resources that are not
+	 *                    in sync with the local file system
+	 * @param derived     Specifying this flag is equivalent to atomically calling
+	 *                    {@link IResource#setDerived(boolean)} immediately after
+	 *                    creating the resource or non-atomically setting the
+	 *                    derived flag before setting the content of an already
+	 *                    existing file
+	 * @param keepHistory a flag indicating whether or not store the current
+	 *                    contents in the local history if the file did already
+	 *                    exist
+	 * @param monitor     a progress monitor, or <code>null</code> if progress
+	 *                    reporting is not desired
+	 * @throws CoreException if this method fails or is canceled.
+	 * @since 3.21
+	 */
+	public default void createOrUpdate(byte[] content, boolean force, boolean derived, boolean keepHistory,
```

Comment:
Maybe instead of a new method one could think about having a new `FLAG` instead like `IResource.CREATE_MISSING` that way it could be used with any of the existing methods as well and one don't need to add more methods for different variations. This is similar to https://docs.oracle.com/javase/8/docs/api/java/nio/file/StandardCopyOption.html and https://docs.oracle.com/javase/8/docs/api/java/nio/file/StandardOpenOption.html in the JDK.

Reply:
Yea probably that would be possible - somehow. But the details are unclear:
Could that flag be used in create() or setContents() or both? 
If both: where is the desired difference then? The problem to me seems that unlike JDK we already have 2 different methods. Other languages that started with different methods like SQL also came up with explicit new functions like 'UPSERT' or 'CREATE OR REPLACE'.
So unless you come up with a clean solution i would prefer to merge as is.

Reply:
I gave it a try to allow the existing(!) IResource.REPLACE flag for create() as implementation. The outcome is that it is possible with minor changes. The implementation does not become simpler, since the main changes come form allowing a byte array.  It would be a little bit incompatible if create() would suddenly accept flags it never rejected and behave different - but not break the documented API. I don't care much which API we want to use. The change would look like this (on top of this PR):
```
diff --git a/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/localstore/FileSystemResourceManager.java b/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/localstore/FileSystemResourceManager.java
index 5cc21b1..8a94d8a 100644
--- a/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/localstore/FileSystemResourceManager.java
+++ b/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/localstore/FileSystemResourceManager.java
@@ -1275,7 +1275,7 @@
 					throw new ResourceException(IResourceStatus.NOT_FOUND_LOCAL, target.getFullPath(), message, null);
 				}
 			} else {
-				if (fileInfo.exists()) {
+				if (!BitMask.isSet(updateFlags, IResource.REPLACE) && fileInfo.exists()) {
 					String message = NLS.bind(Messages.localstore_resourceExists, target.getFullPath());
 					throw new ResourceException(IResourceStatus.EXISTS_LOCAL, target.getFullPath(), message, null);
 				}
diff --git a/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/File.java b/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/File.java
index 5bf8598..c699cae 100644
--- a/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/File.java
+++ b/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/File.java
@@ -126,7 +126,7 @@
 			final ISchedulingRule rule = workspace.getRuleFactory().createRule(this);
 			try {
 				workspace.prepareOperation(rule, subMonitor.newChild(1));
-				checkCreatable();
+				checkCreatable(updateFlags);
 				workspace.beginOperation(true);
 				IFileStore store = getStore();
 				IFileInfo localInfo = create(updateFlags, subMonitor.newChild(40), store);
@@ -172,7 +172,7 @@
 			final ISchedulingRule rule = workspace.getRuleFactory().createRule(this);
 			try {
 				workspace.prepareOperation(rule, subMonitor.newChild(1));
-				checkCreatable();
+				checkCreatable(updateFlags);
 				workspace.beginOperation(true);
 				IFileStore store = getStore();
 				IFileInfo localInfo = create(updateFlags, subMonitor.newChild(40), store);
@@ -202,8 +202,10 @@
 		}
 	}
 
-	private void checkCreatable() throws CoreException {
-		checkDoesNotExist();
+	private void checkCreatable(int updateFlags) throws CoreException {
+		if ((updateFlags & IResource.REPLACE) == 0) {
+			checkDoesNotExist();
+		}
 		Container parent = (Container) getParent();
 		ResourceInfo info = parent.getResourceInfo(false, false);
 		parent.checkAccessible(getFlags(info));
@@ -230,7 +232,7 @@
 				}
 			}
 		} else {
-			if (localInfo.exists()) {
+			if (!BitMask.isSet(updateFlags, IResource.REPLACE) && localInfo.exists()) {
 				// return an appropriate error message for case variant collisions
 				if (!Workspace.caseSensitive) {
 					String name = getLocalManager().getLocalName(store);
diff --git a/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/Workspace.java b/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/Workspace.java
index 3e1dcac..05c7a6c 100644
--- a/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/Workspace.java
+++ b/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/Workspace.java
@@ -60,6 +60,7 @@
 import org.eclipse.core.internal.refresh.RefreshManager;
 import org.eclipse.core.internal.resources.ComputeProjectOrder.Digraph;
 import org.eclipse.core.internal.resources.ComputeProjectOrder.VertexOrder;
+import org.eclipse.core.internal.utils.BitMask;
 import org.eclipse.core.internal.utils.Messages;
 import org.eclipse.core.internal.utils.Policy;
 import org.eclipse.core.internal.utils.StringPoolJob;
@@ -1392,7 +1393,7 @@
 	 * be immediately made derived, hidden and/or team private
 	 */
 	public ResourceInfo createResource(IResource resource, int updateFlags) throws CoreException {
-		ResourceInfo info = createResource(resource, null, false, false, false);
+		ResourceInfo info = createResource(resource, null, false, BitMask.isSet(updateFlags, IResource.REPLACE), false);
 		if ((updateFlags & IResource.DERIVED) != 0)
 			info.set(M_DERIVED);
 		if ((updateFlags & IResource.TEAM_PRIVATE) != 0)
diff --git a/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/IFile.java b/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/IFile.java
index 6910c2d..d4c030e 100644
--- a/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/IFile.java
+++ b/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/IFile.java
@@ -392,6 +392,8 @@
 	 * @param content     the content as byte array
 	 * @param createFlags bit-wise or of flag constants ({@link IResource#FORCE},
 	 *                    {@link IResource#DERIVED}, and
+	 *                    {@link IResource#KEEP_HISTORY}, and
+	 *                    {@link IResource#REPLACE}, and
 	 *                    {@link IResource#TEAM_PRIVATE})
 	 * @param monitor     a progress monitor, or <code>null</code> if progress
 	 *                    reporting is not desired
@@ -428,92 +430,100 @@
 	public default void createOrUpdate(byte[] content, boolean force, boolean derived, boolean keepHistory,
 			IProgressMonitor monitor) throws CoreException {
 		Objects.requireNonNull(content);
-		if (exists()) {
-			if (derived != isDerived()) {
-				setDerived(derived, null);
-			}
-			setContents(content, force, keepHistory, monitor);
-		} else {
-			create(content, force, derived, monitor);
-		}
+		create(content, (force ? IResource.FORCE : 0) | (derived ? IResource.DERIVED : 0)
+				| (keepHistory ? IResource.KEEP_HISTORY : 0) | IResource.REPLACE, monitor);
 	}
 
 	/**
-	 * Creates a new file resource as a member of this handle's parent resource.
-	 * The file's contents will be located in the file specified by the given
-	 * file system path.  The given path must be either an absolute file system
-	 * path, or a relative path whose first segment is the name of a workspace path
-	 * variable.
+	 * Creates a new file resource as a member of this handle's parent resource. The
+	 * file's contents will be located in the file specified by the given file
+	 * system path. The given path must be either an absolute file system path, or a
+	 * relative path whose first segment is the name of a workspace path variable.
 	 * <p>
 	 * The {@link IResource#ALLOW_MISSING_LOCAL} update flag controls how this
 	 * method deals with cases where the local file system file to be linked does
 	 * not exist, or is relative to a workspace path variable that is not defined.
-	 * If {@link IResource#ALLOW_MISSING_LOCAL} is specified, the operation will succeed
-	 * even if the local file is missing, or the path is relative to an undefined
-	 * variable. If {@link IResource#ALLOW_MISSING_LOCAL} is not specified, the operation
-	 * will fail in the case where the local file system file does not exist or the
-	 * path is relative to an undefined variable.
+	 * If {@link IResource#ALLOW_MISSING_LOCAL} is specified, the operation will
+	 * succeed even if the local file is missing, or the path is relative to an
+	 * undefined variable. If {@link IResource#ALLOW_MISSING_LOCAL} is not
+	 * specified, the operation will fail in the case where the local file system
+	 * file does not exist or the path is relative to an undefined variable.
 	 * </p>
 	 * <p>
-	 * The {@link IResource#REPLACE} update flag controls how this
-	 * method deals with cases where a resource of the same name as the
-	 * prospective link already exists. If {@link IResource#REPLACE}
-	 * is specified, then the existing linked resource's location is replaced
-	 * by localLocation's value.  This does <b>not</b>
-	 * cause the underlying file system contents of that resource to be deleted.
-	 * If {@link IResource#REPLACE} is not specified, this method will
-	 * fail if an existing resource exists of the same name.
+	 * The {@link IResource#REPLACE} update flag controls how this method deals with
+	 * cases where a resource of the same name as the prospective link already
+	 * exists. If {@link IResource#REPLACE} is specified, then the existing linked
+	 * resource's location is replaced by localLocation's value. This does
+	 * <b>not</b> cause the underlying file system contents of that resource to be
+	 * deleted. If {@link IResource#REPLACE} is not specified, this method will fail
+	 * if an existing resource exists of the same name.
 	 * </p>
 	 * <p>
-	 * The {@link IResource#HIDDEN} update flag indicates that this resource
-	 * should immediately be set as a hidden resource.  Specifying this flag
-	 * is equivalent to atomically calling {@link IResource#setHidden(boolean)}
-	 * with a value of <code>true</code> immediately after creating the resource.
+	 * The {@link IResource#HIDDEN} update flag indicates that this resource should
+	 * immediately be set as a hidden resource. Specifying this flag is equivalent
+	 * to atomically calling {@link IResource#setHidden(boolean)} with a value of
+	 * <code>true</code> immediately after creating the resource.
 	 * </p>
 	 * <p>
 	 * Update flags other than those listed above are ignored.
 	 * </p>
 	 * <p>
-	 * This method synchronizes this resource with the local file system at the given
-	 * location.
+	 * This method synchronizes this resource with the local file system at the
+	 * given location.
 	 * </p>
 	 * <p>
-	 * This method changes resources; these changes will be reported
-	 * in a subsequent resource change event, including an indication
-	 * that the file has been added to its parent.
+	 * This method changes resources; these changes will be reported in a subsequent
+	 * resource change event, including an indication that the file has been added
+	 * to its parent.
 	 * </p>
 	 * <p>
-	 * This method is long-running; progress and cancellation are provided
-	 * by the given progress monitor.
+	 * This method is long-running; progress and cancellation are provided by the
+	 * given progress monitor.
 	 * </p>
 	 *
 	 * @param localLocation a file system path where the file should be linked
-	 * @param updateFlags bit-wise or of update flag constants
-	 *   ({@link IResource#ALLOW_MISSING_LOCAL}, {@link IResource#REPLACE} and {@link IResource#HIDDEN})
-	 * @param monitor a progress monitor, or <code>null</code> if progress
-	 *    reporting is not desired
-	 * @exception CoreException if this method fails. Reasons include:
-	 * <ul>
-	 * <li> This resource already exists in the workspace.</li>
-	 * <li> The workspace contains a resource of a different type
-	 *      at the same path as this resource.</li>
-	 * <li> The parent of this resource does not exist.</li>
-	 * <li> The parent of this resource is not an open project</li>
-	 * <li> The name of this resource is not valid (according to
-	 *    <code>IWorkspace.validateName</code>).</li>
-	 * <li> The corresponding location in the local file system does not exist, or
-	 * is relative to an undefined variable, and <code>ALLOW_MISSING_LOCAL</code> is
-	 * not specified.</li>
-	 * <li> The corresponding location in the local file system is occupied
-	 *    by a directory (as opposed to a file).</li>
-	 * <li> Resource changes are disallowed during certain types of resource change
-	 *       event notification.  See <code>IResourceChangeEvent</code> for more details.</li>
-	 * <li>The team provider for the project which contains this folder does not permit
-	 *       linked resources.</li>
-	 * <li>This folder's project contains a nature which does not permit linked resources.</li>
-	 * </ul>
+	 * @param updateFlags   bit-wise or of update flag constants
+	 *                      ({@link IResource#ALLOW_MISSING_LOCAL},
+	 *                      {@link IResource#REPLACE} and {@link IResource#HIDDEN})
+	 * @param monitor       a progress monitor, or <code>null</code> if progress
+	 *                      reporting is not desired
+	 * @exception CoreException              if this method fails. Reasons include:
+	 *                                       <ul>
+	 *                                       <li>This resource already exists in the
+	 *                                       workspace.</li>
+	 *                                       <li>The workspace contains a resource
+	 *                                       of a different type at the same path as
+	 *                                       this resource.</li>
+	 *                                       <li>The parent of this resource does
+	 *                                       not exist.</li>
+	 *                                       <li>The parent of this resource is not
+	 *                                       an open project</li>
+	 *                                       <li>The name of this resource is not
+	 *                                       valid (according to
+	 *                                       <code>IWorkspace.validateName</code>).</li>
+	 *                                       <li>The corresponding location in the
+	 *                                       local file system does not exist, or is
+	 *                                       relative to an undefined variable, and
+	 *                                       <code>ALLOW_MISSING_LOCAL</code> is not
+	 *                                       specified.</li>
+	 *                                       <li>The corresponding location in the
+	 *                                       local file system is occupied by a
+	 *                                       directory (as opposed to a file).</li>
+	 *                                       <li>Resource changes are disallowed
+	 *                                       during certain types of resource change
+	 *                                       event notification. See
+	 *                                       <code>IResourceChangeEvent</code> for
+	 *                                       more details.</li>
+	 *                                       <li>The team provider for the project
+	 *                                       which contains this folder does not
+	 *                                       permit linked resources.</li>
+	 *                                       <li>This folder's project contains a
+	 *                                       nature which does not permit linked
+	 *                                       resources.</li>
+	 *                                       </ul>
 	 * @exception OperationCanceledException if the operation is canceled.
-	 * Cancelation can occur even if no progress monitor is provided.
+	 *                                       Cancelation can occur even if no
+	 *                                       progress monitor is provided.
 	 * @see IResource#isLinked()
 	 * @see IResource#ALLOW_MISSING_LOCAL
 	 * @see IResource#REPLACE

```

Reply:
I updated the implementation to use IResource#REPLACE and renamed the new API to 'createOrReplace'.

Reply:
What do you think about naming it just `write()`, similar like `Files.write()` from the JDK?
Semantically this could include both cases and the flag would also fit.

Reply:
I agree `doXOrDoY` API doesn't look so nice and efficient, so +1 for `write` or a variation of `setContents`. I'm also under the impression that the new `create` method with a boolean for derived could probably have been spared, as it increases the size of the API without adding new capability here. And the purpose of flags is exactly to prevent from growing the API methods for every new option.
In a comment yesterday, you mention allowing  `create` to allow the `REPLACE` flag to not fail in case resource is already existing. This seems like a good idea, why was it abandoned? That would require less new methods (only tweaking the existing `create` and adding a new one taking a byte[] as input instead of a stream).

Reply:
> you mention allowing `create` to allow the `REPLACE` flag to not fail in case resource is already existing. This seems like a good idea, why was it abandoned?

create does allow REPLACE now. But i would not use that for clients, as it does not give a warning that the resource bundle may not may support it, as it is only behaviour change but not new semanitc API.

> Files.write()

ok for me. i will create PR to rename it.

Reply:
>  as it does not give a warning that the resource bundle may not support it, as it is only behaviour change but not new semanitc API.

Adding a new flag is an API change that requires a version increment so everyone that use that requires to raise their minimal version, I'm not sure what kind of "warning" is meant here?

Reply:
It's not new, it just has an additional meaning/use-case:

![image](https://github.com/eclipse-platform/eclipse.platform/assets/208716/b84c03ff-620e-414f-b282-65d57fc408ad)